File: Data Structures & Algorithms/count-number-of-islands/submission-0.java
class Solution {
    public int numIslands(char[][] grid) {
        int res = 0;
        int[][] visited = new int[grid.length][grid[0].length];

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j <grid[i].length; j++) {
                if (grid[i][j] == '1') {
                    res += dfs(i, j, grid, visited);
                }
            }
        } 

        return res;
    }

    private int dfs(int r, int c, char[][] grid, int[][] visited) {
        if (r < 0 || r >= grid.length || c < 0 || c >= grid[0].length
            || grid[r][c] != '1' || visited[r][c] == 1) {
            return 0;
        }

        visited[r][c] = 1;

        dfs(r + 1, c, grid, visited);
        dfs(r - 1, c, grid, visited);
        dfs(r, c + 1, grid, visited);
        dfs(r, c - 1, grid, visited);

        return 1;
    }
}

File: Data Structures & Algorithms/combination-target-sum-ii/submission-0.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        subset = []
        candidates.sort()

        def dfs(i, curr_sum):
            if curr_sum == target:
                res.append(subset.copy())
                return
            
            if i >= len(candidates) or curr_sum > target:
                return

            subset.append(candidates[i])
            dfs(i + 1, curr_sum + candidates[i])

            subset.pop()
            # escape all the equal numbers
            while i + 1 < len(candidates) and candidates[i] == candidates[i + 1]:
                i += 1
            dfs(i + 1, curr_sum)

        dfs(0, 0)
        return res